### PR TITLE
Change unix domain socket file to be ownned by current user #54

### PR DESCRIPTION
**Pull request by duanbing**

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
@vdemeester ptal?

**Comment by duanbing** [comment]
@vdemeester  Your suggestion is great!  At the beginning I also have not idea why I should pass PID and UID, why not share those with the caller process by default?  If so, I would like to do another updating.  What's more,  docker/pkg/plugins and some other pkg related also need to update.

**Comment by vdemeester** [comment]
@duanbing yes except keeping `NewUnixSocket` as now (because it's kinda widely used) and have that func named `NewUnixSocketWithOpt`.

```go
// +build !windows

package sockets

import (
    "net"
    "os"
    "syscall"
)

// Usage:
//     l, err := sockets.NewUnixSocket("/path/to/sockets",sockets.WithChown(0,0),sockets.WithChmod(0660))

type SockOption func(string) error

func WithChown(uid, gid int) SockOption {
    return func(path string) error {
        if err := os.Chown(path, uid, gid); err != nil {
            return err
        }
        return nil
    }
}

func WithChmod(mask int) SockOption {
    return func(path string) error {
        if err := os.Chmod(path, mask); err != nil {
            return err
        }
        return nil
    }
}

// NewUnixSocketWithOpt creates a unix socket with the specified options.
func NewUnixSocketWithOpt(path string, opts ...func(string) error) (net.Listener, error) {
    if err := syscall.Unlink(path); err != nil && !os.IsNotExist(err) {
        return nil, err
    }
    mask := syscall.Umask(0777)
     defer syscall.Umask(mask)

    l, err := net.Listen("unix", path)
    if err != nil {
        return nil, err
    }

    for _, op := range opts {
        if err := op(path); err != nil {
            l.Close()
            return nil, err
        }
    }

    return l, nil
}
// NewUnixSocket creates a unix socket with the specified path and group.
func NewUnixSocket(path string, uid int) (net.Listener, error) {
	return NewUnixSocketWithOpt(path, WithChown(uid, 0), WithChmod(0777))
}
```

**Comment by duanbing** [comment]
@vdemeester PTAL.

**Comment by duanbing** [comment]
@vdemeester  Comments had added.

File: sockets/unix_socket.go [hunk 1]
```diff
@@ -1,5 +1,51 @@
 // +build !windows
 
+/*
+Package sockets is a simple unix domain socket wrapper.
```
Comment (vdemeester) [inline sockets/unix_socket.go]: This example could be defined as an example : see https://blog.golang.org/examples. But can be done in a follow-up :wink:

File: sockets/unix_socket.go [hunk 2]
```diff
@@ -20,13 +89,18 @@ func NewUnixSocket(path string, gid int) (net.Listener, error) {
...
+			l.Close()
+			return nil, err
+		}
 	}
+
 	return l, nil
 }
+
+// NewUnixSocket creates a unix socket with the specified path and group.
+func NewUnixSocket(path string, uid int) (net.Listener, error) {
```
Comment (vdemeester) [inline sockets/unix_socket.go]: oh wait, this was the other way around :sweat_smile: 
`NewUnixSocket(path string, gid int)` and `os.Chown(path, 0, gid)`

File: sockets/unix_socket.go (hunk 1 above)
Comment (duanbing) [reply sockets/unix_socket.go]: Ok,  Will do it in a follow-up.

**Pull request merged**